access-grant/src/main/java/com/inrupt/client/accessgrant/AccessCredentialQuery.java

```diff
@@ -0,0 +1,228 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.accessgrant;
+
+import java.net.URI;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * An object to represent an access credential query.
+ *
+ * @param <T> The access credential type
+ */
+public class AccessCredentialQuery<T extends AccessCredential> {
+
+    private static final URI SOLID_ACCESS_GRANT = URI.create("SolidAccessGrant");
+    private static final URI SOLID_ACCESS_REQUEST = URI.create("SolidAccessRequest");
+    private static final URI SOLID_ACCESS_DENIAL = URI.create("SolidAccessDenial");
+
+    private final URI type;
+    private final Set<URI> purposes;
+    private final Set<String> modes;
+    private final URI resource;
+    private final URI creator;
+    private final URI recipient;
+    private final Class<T> clazz;
+
+    /**
+     * Create an access credential query.
+     *
+     * @param resource the resource, may be {@code null}
+     * @param creator the creator, may be {@code null}
+     * @param recipient the recipient, may be {@code null}
+     * @param purposes the purposes, never {@code null}
+     * @param modes the access modes, never {@code null}
+     * @param clazz the credential type, never {@code null}
+     */
+    AccessCredentialQuery(final URI resource, final URI creator, final URI recipient,
+            final Set<URI> purposes, final Set<String> modes, final Class<T> clazz) {
+        this.clazz = Objects.requireNonNull(clazz, "The clazz parameter must not be null!");
+
+        if (AccessGrant.class.isAssignableFrom(clazz)) {
+            this.type = SOLID_ACCESS_GRANT;
+        } else if (AccessRequest.class.isAssignableFrom(clazz)) {
+            this.type = SOLID_ACCESS_REQUEST;
+        } else if (AccessDenial.class.isAssignableFrom(clazz)) {
+            this.type = SOLID_ACCESS_DENIAL;
+        } else {
+            throw new AccessGrantException("Unsupported type " + clazz + " in query request");
+        }
+
+        this.resource = resource;
+        this.creator = creator;
+        this.recipient = recipient;
+        this.purposes = purposes;
+        this.modes = modes;
+    }
+
+    /**
+     * Get the access credential type value.
+     *
+     * @return the type, never {@code null}
+     */
+    public URI getType() {
+        return type;
+    }
+
+    /**
+     * Get the requested resource.
+     *
+     * @return the resource, may be {@code null}
+     */
+    public URI getResource() {
+        return resource;
```

Comment:
Even if it may not be supported by all Access Management Applications, an Access Grant may cover multiple resources. Is it something that should be reflected in the querying capabilities? Or do we want to push this constraint down to the client libraries, in which case we should consider the same for the JC client.